src/app/Library/CrudPanel/Traits/MorphRelationships.php

```diff
@@ -0,0 +1,210 @@
+            [$morphTypeField, $morphIdField] = $field['subfields'];
+            Arr::set($input, $morphTypeField['name'], Arr::get($input, $field['name'].'.'.$morphTypeField['name']));
+            Arr::set($input, $morphIdField['name'], Arr::get($input, $field['name'].'.'.$morphIdField['name']));
+        }
+
+        return $input;
+    }
+
+    /**
+     * This function created the MorphTo relation fields in the CrudPanel.
+     *
+     * @param  array  $field
+     * @return void
+     */
+    private function createMorphToRelationFields(array $field, $morphTypeFieldName, $morphIdFieldName)
+    {
+        $morphTypeField = static::getMorphTypeFieldStructure($field['name'], $morphTypeFieldName);
+        $morphIdField = static::getMorphIdFieldStructure($field['name'], $morphIdFieldName, $morphTypeFieldName);
+        $morphIdField['morphMap'] = $morphTypeField['morphMap'] = (new $this->model)->{$field['name']}()->morphMap();
+        $field['subfields'] = [$morphTypeField, $morphIdField];
+
+        return $field;
+    }
+
+    /**
+     * Return the relation field names for a morphTo field.
+     *
+     * @param  string  $relationName  the morphto relation name
+     * @return array
+     */
+    private function getMorphToFieldNames(string $relationName)
+    {
+        $relation = (new $this->model)->{$relationName}();
+
+        return [$relation->getMorphType(), $relation->getForeignKeyName()];
+    }
+
+    /**
+     * Make sure morph fields have the correct structure.
+     *
+     * @param  array  $field
+     * @return array
+     */
+    private function makeSureMorphSubfieldsAreDefined(array $field)
+    {
+        if (isset($field['relation_type']) && $field['relation_type'] === 'MorphTo') {
+            [$morphTypeFieldName, $morphIdFieldName] = $this->getMorphToFieldNames($field['name']);
+            if (! $this->hasFieldWhere('name', $morphTypeFieldName) || ! $this->hasFieldWhere('name', $morphIdFieldName)) {
+                // create the morph fields in the crud panel
+                $field = $this->createMorphToRelationFields($field, $morphTypeFieldName, $morphIdFieldName);
+                foreach ($field['morphOptions'] ?? [] as $morphOption) {
+                    [$key, $label, $options] = $this->getMorphOptionStructured($morphOption);
+                    $field = $this->addMorphOption($field, $key, $label, $options);
+                }
+            }
+        }
+
+        return $field;
+    }
+
+    /**
+     * This function is responsible for setting up the morph fields structure.
+     * Developer can define the morph structure as follows:
+     *  'morphOptions => [
+     *       ['nameOnAMorphMap', 'label', [options]],
+     *       ['App\Models\Model'], // display the name of the model
+     *       ['App\Models\Model', 'label', ['data_source' => backpack_url('smt')]
+     *  ]
+     * OR
+     * ->addMorphOption('App\Models\Model', 'label', ['data_source' => backpack_url('smt')]).
```

Comment:
I like this syntax explained in the comments, but it's confusing to have it here. That's not the syntax for THIS method, it's the syntax for the same method on the CrudField object. I only tracked that down after wanting to point out this method's signature has one more parameter 😀

So let's move this comment there please. And write out for each parameter what it actually is.